libpod/pod.go

```diff
@@ -71,19 +72,167 @@ func (p *Pod) Init() error {
+// starting containers and immediately report an error
+// Start() is not an atomic operation - if an error is reported, containers that
+// have already started will remain running
 func (p *Pod) Start() error {
-	return ErrNotImplemented
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	if !p.valid {
+		return ErrPodRemoved
+	}
+
+	allCtrs, err := p.runtime.state.PodContainers(p)
+	if err != nil {
+		return err
+	}
+
+	// We need to lock all the containers
+	for _, ctr := range allCtrs {
+		ctr.lock.Lock()
+		defer ctr.lock.Unlock()
+
+		if err := ctr.syncContainer(); err != nil {
+			return err
+		}
+	}
+
+	// Send a signal to all containers
+	for _, ctr := range allCtrs {
+		// Ignore containers that are not created or stopped
+		if ctr.state.State != ContainerStateCreated && ctr.state.State != ContainerStateStopped {
+			continue
+		}
+
+		// TODO remove this when we patch conmon to support restarting containers
+		if ctr.state.State == ContainerStateStopped {
+			continue
+		}
+
+		if err := ctr.runtime.ociRuntime.startContainer(ctr); err != nil {
+			return errors.Wrapf(err, "error starting container %s", ctr.ID())
+		}
+
+		// We can safely assume the container is running
+		ctr.state.State = ContainerStateRunning
+
+		if err := ctr.save(); err != nil {
+			return err
+		}
+
+		logrus.Debugf("Started container %s", ctr.ID())
+	}
+
+	return nil
 }
 
 // Stop stops all containers within a pod that are not already stopped
+// Each container will use its own stop timeout
+// Only running containers will be stopped. Paused, stopped, or created
+// containers will be ignored.
+// If an error is encountered stopping any one container, no further containers
+// will be stopped, and an error will immediately be returned.
+// Stop() is not an atomic operation - if an error is encountered, containers
+// which have already been stopped will not be restarted
 func (p *Pod) Stop() error {
-	return ErrNotImplemented
+	p.lock.Lock()
+	defer p.lock.Unlock()
+
+	if !p.valid {
+		return ErrPodRemoved
+	}
+
+	allCtrs, err := p.runtime.state.PodContainers(p)
+	if err != nil {
+		return err
+	}
+
+	// We need to lock all the containers
+	for _, ctr := range allCtrs {
+		ctr.lock.Lock()
+		defer ctr.lock.Unlock()
+
+		if err := ctr.syncContainer(); err != nil {
+			return err
+		}
+	}
+
+	// Send a signal to all containers
+	for _, ctr := range allCtrs {
+		// Ignore containers that are not running
+		if ctr.state.State != ContainerStateRunning {
+			continue
+		}
+
+		if err := ctr.runtime.ociRuntime.stopContainer(ctr, ctr.config.StopTimeout); err != nil {
+			return errors.Wrapf(err, "error stopping container %s", ctr.ID())
+		}
+
+		// Sync container state to pick up return code
+		if err := ctr.runtime.ociRuntime.updateContainerStatus(ctr); err != nil {
+			return err
+		}
+
+		// Clean up storage to ensure we don't leave dangling mounts
+		if err := ctr.cleanupStorage(); err != nil {
+			return err
+		}
+
+		logrus.Debugf("Stopped container %s", ctr.ID())
```

Comment:
I think this messages should be moved up to the point when the container is stopped, so it is clear even if anything fails that this container was stopped.